glusterd2/commands/snapshot/snapshot-restore.go

```diff
@@ -0,0 +1,375 @@
+	if err != nil {
+		return err
+	}
+	offlineBricks, err := snapshot.GetOfflineBricks(snapVol)
+	if err != nil {
+		return err
+	}
+
+	//Do a proper snapshot stop, once there is a generic way of stopping all the proceess of a volume.
+	for _, b := range onlineBricks {
+		if err = b.TerminateBrick(); err != nil {
+			if err = b.StopBrick(c.Logger()); err != nil {
+				return err
+			}
+		}
+		if err := unix.Setxattr(b.Path, volumeIDXattrKey, []byte(volinfo.ID), unix.XATTR_REPLACE); err != nil {
+			return err
+		}
+	}
+
+	for _, b := range offlineBricks {
+		if err := snapshot.MountSnapBrickDirectory(snapVol, &b); err != nil {
+			return err
+		}
+		if err := unix.Setxattr(b.Path, volumeIDXattrKey, []byte(volinfo.ID), unix.XATTR_REPLACE); err != nil {
+			return err
+		}
+	}
+
+	//TODO Stop other process of snapshot volume
+
+	return nil
+}
+
+func remountBrick(b brick.Brickinfo, volinfo *volume.Volinfo) error {
+	if err := snapshot.UmountBrick(b); err != nil {
+		return err
+	}
+	err := snapshot.MountSnapBrickDirectory(volinfo, &b)
+	return err
+
+}
+
+func undoSnapStore(c transaction.TxnCtx) error {
+	var snapInfo snapshot.Snapinfo
+	var volinfo volume.Volinfo
+
+	if err := c.Get("snapinfo", &snapInfo); err != nil {
+		return err
+	}
+
+	if err := c.Get("volinfo", &volinfo); err != nil {
+		return err
+	}
+
+	if err := snapshot.AddOrUpdateSnapFunc(&snapInfo); err != nil {
+		c.Logger().WithError(err).WithField(
+			"volume", volinfo.Name).Debug("storeSnapshot: failed to store snapshot info")
+		return err
+	}
+
+	if err := volume.AddOrUpdateVolumeFunc(&volinfo); err != nil {
+		c.Logger().WithError(err).WithField(
+			"volume", volinfo.Name).Debug("failed to store volume info")
+		return err
+	}
+
+	if err := volgen.Generate(); err != nil {
+		c.Logger().WithError(err).WithField(
+			"volume", volinfo.Name).Debug("failed to generate volfiles")
+		return err
+	}
+
+	return nil
+}
+
+func undoSnapRestore(c transaction.TxnCtx) error {
+	var snapname string
+	if err := c.Get("snapname", &snapname); err != nil {
+		return err
+	}
+
+	snapInfo, err := snapshot.GetSnapshot(snapname)
+	if err != nil {
+		return err
+	}
+	snapVol := &snapInfo.SnapVolinfo
+
+	for _, b := range snapVol.GetLocalBricks() {
+
+		if err := remountBrick(b, snapVol); err != nil {
+			return err
+		}
+		if snapVol.State == volume.VolStarted {
+			if err := b.StartBrick(c.Logger()); err != nil {
+				if err == errors.ErrProcessAlreadyRunning {
+					continue
+				}
+				return err
+			}
+		} else {
+			if err = b.TerminateBrick(); err != nil {
+				if err = b.StopBrick(c.Logger()); err != nil {
+					//Process might not be running,
+					//TODO once we have errors.ErrProcessAlreadyStopped
+					//check for other errors
+				}
+				if err := snapshot.UmountBrick(b); err != nil {
+					return err
+				}
+			}
+
+		}
+	}
+
+	return nil
+}
+
+func createVolumeBrickFromSnap(bricks []brick.Brickinfo, vol *volume.Volinfo) []brick.Brickinfo {
+	var newBricks []brick.Brickinfo
```

Comment:
suggestion: append may consume more memory (if the length exceeds it increments the capacity by twice ), I would suggest making slice capacity as a length of brickstatus

newBricks =make ([]brick.Brickinfo, 0, len(bricks)

Reply:
same as above command